src/TabContainer.js

```diff
@@ -1,138 +1,134 @@
-     */
-    activeKey: PropTypes.any,
-  };
-
-  constructor(props) {
-    super(props);
-
-    this.state = {
-      tabContext: {
-        onSelect: this.props.onSelect,
-        activeKey: this.props.activeKey,
-        transition: this.props.transition,
-        mountOnEnter: this.props.mountOnEnter,
-        unmountOnExit: this.props.unmountOnExit,
-        getControlledId: this.getControlledId,
-        getControllerId: this.getControllerId,
-      },
-    };
-  }
-
-  static getDerivedStateFromProps(
-    { activeKey, mountOnEnter, unmountOnExit, transition },
-    prevState,
-  ) {
-    return {
-      tabContext: {
-        ...prevState.tabContext,
-        activeKey,
-        mountOnEnter,
-        unmountOnExit,
-        transition,
-      },
-    };
-  }
-
-  getKey(key, type) {
-    const { generateChildId, id } = this.props;
-    if (generateChildId) return generateChildId(key, type);
-    return id ? `${id}-${type}-${key}` : null;
-  }
-
-  getControlledId = key => this.getKey(key, 'tabpane');
-
-  getControllerId = key => this.getKey(key, 'tab');
-
-  render() {
-    const { children, onSelect } = this.props;
-
-    return (
-      <TabContext.Provider value={this.state.tabContext}>
-        <SelectableContext.Provider value={onSelect}>
-          {children}
-        </SelectableContext.Provider>
-      </TabContext.Provider>
-    );
-  }
-}
-
-export default uncontrollable(TabContainer, { activeKey: 'onSelect' });
+    }
+
+    return error;
+  },
+
+  /**
+   * Sets a default animation strategy for all children `<TabPane>`s. Use
+   * `false` to disable, `true` to enable the default `<Fade>` animation or
+   * a react-transition-group v2 `<Transition/>` component.
+   *
+   * @type {{Transition | false}}
+   * @default {Fade}
+   */
+  transition: PropTypes.oneOfType([
+    PropTypes.oneOf([false]),
+    PropTypes.elementType,
+  ]),
+  /**
+   * Wait until the first "enter" transition to mount tabs (add them to the DOM)
+   */
+  mountOnEnter: PropTypes.bool,
+
+  /**
+   * Unmount tabs (remove it from the DOM) when they are no longer visible
+   */
+  unmountOnExit: PropTypes.bool,
+
+  /**
+   * A function that takes an `eventKey` and `type` and returns a unique id for
+   * child tab `<NavItem>`s and `<TabPane>`s. The function _must_ be a pure
+   * function, meaning it should always return the _same_ id for the same set
+   * of inputs. The default value requires that an `id` to be set for the
+   * `<TabContainer>`.
+   *
+   * The `type` argument will either be `"tab"` or `"pane"`.
+   *
+   * @defaultValue (eventKey, type) => `${props.id}-${type}-${eventKey}`
+   */
+  generateChildId: PropTypes.func,
+
+  /**
+   * A callback fired when a tab is selected.
+   *
+   * @controllable activeKey
+   */
+  onSelect: PropTypes.func,
+
+  /**
+   * The `eventKey` of the currently active tab.
+   *
+   * @controllable onSelect
+   */
+  activeKey: PropTypes.any,
+};
+
+const TabContainer = props => {
```

Comment:
We should be using `React.forwardRef` here (#4194).

Reply:
Does this actually render a DOM node though?

Reply:
If https://github.com/reactjs/rfcs/pull/97 lands, I think the children could be wrapped in a fragment with ref, but that proposal is still open.